### DISCUSSION TITLE
Suggestion for an easier to use image component

### DISCUSSION THREAD

**Discussion by Princesseuh** [withastro/roadmap#447] [Proposal]
Proof of concept implementation PR of this proposal: https://github.com/withastro/astro/pull/5943

# Summary

According to our users, the image integration is **CCC**: Cumbersome, confusing and complicated to use. It's unclear for users why certain props are required, how the different properties impact the resulting image(s) or how the background behavior can differs massively depending on seemingly inoffensive changes to their code.

This proposal aims to outline a plan to make the integration overall easier to use and easier to understand

# Background & Motivation

We very often get issues and support threads about the image integration, users claiming it to be buggy when many times, it just turns out that the user misunderstood how to achieve the specific result they were aiming for.

While we could make the current behavior more clear through better error messages, better documentation, better types etc, and it would massively improve the experience already; I believe we can also improve the core experience through the following changes:

## Easier nomenclature from the perspective of the user

While it makes a lot of sense from the technical perspective, from the point of view of the user, it is unclear why an image they'd refer to it like such: `<Image src="../assets/my_image.png" />` would be considered to be "remote". As such, I propose the following changes:

- Local images are either **statically imported** or **dynamically referred** to. The former meaning any ESM-imported images, whereas the latter refers to any raw string references to a local file.
- Remote images exclusively refers to images that requires a fetch request to an external host.
    - Despite technically needing a fetch, in SSR, a dynamic reference to a file in `public` should be considered a "local image"

## Reduce the number of needed props in the more common cases

> 💡 From now on, the previously suggested nomenclature will be used

For statically imported images, nothing changes. None of the props were required in the first place.

Currently, for dynamically referred local images, `width`, `height` (and `aspectRatio`, if both were not specified), and `format` are required props. 

- In SSG, dynamically referred local images are available to us to read, as such it is possible to infer `width`, `height` (and as such `aspectRatio`), and `format` from them, removing the need for user to manually pass those themselves. 
- In SSR, the same goes, however this would only apply to images from the public folder, as `src` isn't available at runtime. (Referring to an image from `src` in SSR should probably be an error pending a better assets story in Astro)

For remote images the same thing applies:  `width`, `height` (and `aspectRatio`, if both were not specified), and `format` are required props. In both SSG and SSR, it is possible for us to fetch the image and infer all the required props. 

### Performance concerns

Most of this inferring business happens during the page rendering (since we need the inferred data to properly fill the `width` and `height` attributes to avoid CLS), this can slow down the page rendering. In both SSG and SSR, this is not a real problem for either statically imported or dynamically referred files, if reading a single file from the file system is slow, we probably have bigger problems.

For remote images, this is a more real concern, because it means that to render the page, a `fetch` call has to finish. In SSG, this is annoying, but not a real problem: You need the `fetch` to finish eventually to build the page. In SSR however, this means delaying the page render. 

Possible solutions:
- No inferring for remote images in SSR at all.
- Allow users to opt out of inferring for remote images in SSR.
- Swallow the cost, but use a library like https://github.com/nodeca/probe-image-size to at least attempt to do it as efficiently as possible. (some users are already doing this themselves!)

In all cases (statically, dynamic, SSG and SSR), caching is possible, so ultimately this can delay a single request at worse. 

Double fetching is not a concern, as the image will either be cached in the HTTP cache or through a filesystem-based one.

# Goals

- Make the image integration easier to use, doing so by
- Inferring as many parameters as possible 
- While not losing any points in regards to CLS

# Non-goals of this proposal

In order to keep this proposal simple, the following (understandably commonly-requested) features are not included in this proposal

- How to include optimized images inside non-Astro components
- How to include optimized images inside Markdown
- Placeholders

# Example

Similar behaviour is to be expected in all cases, no matter the loader used or the rendering technique (SSG or SSR). As such, these examples will only cover a single case to showoff behaviour

### **Code**

```jsx
<Image src={"/princesses_public.webp"} height={300} />
```

### Old behaviour

```
Error: "aspectRatio" must be included if only "height" is provided
```

### New behaviour

`width` automatically inferred from source image, results in 700x300, keeping original aspect ratio

**Result HTML**
```html
<img width="700" height="300" src="/_image?f=jpg&amp;w=700&amp;h=300&amp;href=%2Fprincesses_public.webp" loading="lazy" decoding="async">
```

**Comment by tusamni** [comment]
Inferring the width / height for remote images using fetch would be nice upgrade.

Currently I've built my own image component to do that.

**Comment by matthewp** [comment]
@Princesseuh this is a great start. The idea of local/remote does seem to be confusing. Is there an argument to be made not to support "remote" images that are in fact local? Why not use statically imported?

Or is there a use-case for the local-in-public images that I'm not seeing?

**Reply by Princesseuh** [reply]
It's common for people to refer to images dynamically, like so:

```html
<Image
  src={`/images/beers/${beer.image}.png`}
/>
```

(other instances would be image comes from a markdown frontmatter, a JSON files etc)

**Comment by tony-sull** [comment]
> Swallow the cost, but use a library like https://github.com/nodeca/probe-image-size to at least attempt to do it as efficiently as possible. (some users are already doing this themselves!)

I'd be curious to see some performance benchmarks for this kind of approach 👀   It looks like `probe-image-size` has a few tricks up its sleeve to try to fetch only enough of the image binary to read metadata, but I'm not sure what that might look like if an SSR page needs to block on more than a few images

**Comment by gvkhna** [comment]
After spending some time on this component. I think I have an idea of a solution that could help to disambiguate local versus remote more concretely.

Local can be referred to by either of the two following URLs
`file://` or no protocol as in `/images/myimage.png` or `@images/myimage.png`

All others are treated as remote:
`http?s://` even `ftp://` is a valid remote URL I believe. 

I wanted to make this distinction more clear because in the code there's some checking and branching behind the scenes, for trying to determine this, but a rules based approach could make the code simpler. 

Is there any situation where a string without a protocol that isn't `file://` could be local? At least if there is that could be out of scope for the sake of the image component but I don't believe there is.

**Reply by Princesseuh** [reply]
That seems in line with what the proposal here outlines, which is that any string reference that does not requires a fetch to an external host (so, no `http(s)`, `ftp` etc) would be considered a local image. The proposal needs to be slightly more verbose to include ESM-imported images, as they technically behave differently.

To summarize in simpler word:
- ESM import / string that does not start with `http(s)` = Local
- Everything else = Remote

Let me know if that makes sense!

**Comment by beeb** [comment]
I fully agree with the proposal, at the moment it is confusing to require different props depending on how a "local" image is referred to.

Since we are talking about breaking changes in the image integration, I would like to raise another topic however, that I summarized here: https://github.com/withastro/astro/pull/4800#issuecomment-1408538313

I first added the `background` option in https://github.com/withastro/astro/pull/4642 to allow me to generate optimized images (be it webp, avif or jpeg) from a master transparent PNG image. Removing the alpha layer and replacing it with a solid color that matches the context inside the page allows to reduce the file size in general.

However, https://github.com/withastro/astro/pull/4800 changed this behavior (I just noticed it today when upgrading to astro v2) to only perform the background replacement if the format doesn't support alpha layers. This was a breaking change that I would like to revert. Users wanting to keep the alpha layers should simply not use the `background` prop. Usage of said prop indicates intent to replace the background with a solid color.

Whenever appropriate, I would like to propose a change that would look like this to restore the old behavior: https://github.com/beeb/astro/commit/7a78ce6264595ed2c6da3cae2e991d699391fcce

**Reply by beeb** [reply]
To add to this, and because @tony-sull questioned the need for this option, I would argue that having the option to set a background color is needed solong as we can transform the format from one that supports alpha to one that doesn't (e.g. png to jpeg), since otherwise, any default background color might conflict with the contents of the image. Think black logo on transparent background, if black is used when transforming to jpeg, the logo information is lost (and the image might not integrate nicely with its surroundings). See my initial issue on the topic: https://github.com/withastro/astro/issues/4625

**Comment by Princesseuh** [comment]
Superseded by https://github.com/withastro/roadmap/discussions/468

Thank you everyone for sharing your feedback!

**Comment by matthewp** [comment]
This happened.
